components/membership/group-policy-impl/src/main/kotlin/net/corda/membership/impl/grouppolicy/GroupPolicyProviderImpl.kt

```
@@ -151,83 +213,134 @@ class GroupPolicyProviderImpl @Activate constructor(
+        if (vNodeInfo == null) {
+            logger.warn("Could not get virtual node info for holding identity [${holdingIdentity}]")
+        }
+        val metadata = vNodeInfo?.cpiIdentifier?.let { cpiInfoReader.get(it) }
+        if (metadata == null) {
+            logger.warn(
+                "Could not get CPI metadata for holding identity [${holdingIdentity}] and CPI with " +
+                        "identifier [${vNodeInfo?.cpiIdentifier.toString()}]"
+            )
+        }
+        fun persistedPropertyQuery(): LayeredPropertyMap? = try {
+            membershipQueryClient.queryGroupPolicy(holdingIdentity).getOrThrow()
+        } catch (e: MembershipQueryResult.QueryException) {
+            logger.warn("Failed to retrieve persisted group policy properties.", e)
+            null
+        }
+        return try {
+            groupPolicyParser.parse(
+                holdingIdentity,
+                metadata?.groupPolicy,
+                ::persistedPropertyQuery
+            )
+        } catch (e: BadGroupPolicyException) {
+            logger.warn("Failed to parse group policy. Returning null.", e)
+            null
+        }
+    }
+
+    /**
+     * Registers callback when MGM has finished its registration and has the final group policy persisted.
+     * This will make sure we have the trust stores and other important information in the group policy ready.
+     */
+    inner class FinishedRegistrationsProcessor : DurableProcessor<String, MembershipEvent> {
+        @Suppress("NestedBlockDepth")
+        override fun onNext(events: List<Record<String, MembershipEvent>>): List<Record<*, *>> {
+            logger.info("Received event after mgm registration.")
+            events.forEach {record ->
+                try {
+                    record.value
+                        ?: throw CordaRuntimeException("MembershipEvent with record key: ${record.key} was null.")
```

Do we want to throw an exception?

---

I think yes. We are doing this in other services as well. It will be caught and logged out anyway, but if we are not throwing it will try to go through the execution with null values which is far more worse.